### PR TITLE
Allow mixing multi and pipeline modes

### PR DESCRIPTION
**Pull request by yatsukhnenko**
@michael-grunder, this is still WIP but I need your reviewing at this point

### CODE REVIEW

**Comment by michael-grunder** [comment]
Nice,

This is a good feature that will make many people happy.  I like how you're flipping multi on and off with bitwise magic :smiley:

I'll give this a good test over the weekend!

**Comment by yatsukhnenko** [comment]
@michael-grunder, all tests passed ^_^
Did you look at code? I doubt about `exec` command. Should we send it if we are in pipelined multi mode or just enqueue and send after the second exec call?

**Comment by michael-grunder** [comment]
Hey,

I only glanced at the code but unfortunately I can't get it to work.  The tests pass for me too, but I think that's because the tests aren't trying to mix modes.

When I try to execute multi blocks inside a pipeline the code blocks for me, as in the following script:

```php
<?php

$r = new Redis();
$r->connect('127.0.0.1', 6379);

echo "Entering pipeline...\n";
$r->pipeline();

echo "Executing inner multi!\n";
$r = $r->multi();
echo "Executing set1...\n";
$r->set('myset1-1', 'myval1-1');  // <-- blocks here, waiting for a reply to the SET command
echo "Executing set2...\n";
$r->set('myset1-2', 'myval1-2');
echo "Executing inner exec:\n";
var_dump($r->exec());

echo "Executing outer exec:\n";
var_dump($r->exec());
?>
```

I may be missing something though, as I'm operating on too little sleep :)

**Comment by yatsukhnenko** [comment]
Ah, I forgot about little change... Try again, pls

**Comment by yatsukhnenko** [comment]
You can even change order of `pipeline` and `multi` calls and it will be the same result :) But, as you can see there is no inner or outer `exec` and I doubt about this solution

**Comment by yatsukhnenko** [comment]
@michael-grunder, would you have a time to look at this PR once again? I need to know what do you think about single/double exec 🤔

**Comment by ovr** [comment]
@yatsukhnenko Can you rebase your branch? 
ping @michael-grunder 

will be rly cool to have this feature 👍

**Comment by ovr** [comment]
ping @michael-grunder

**Comment by michael-grunder** [comment]
@yatsukhnenko I'm not sure if this is your specific question, but I can see using the following pattern:

```php
/* Enter a pipeline */
$r->pipeline();

/* First multi block, pipelined */
$r->multi();
$r->set('foo1','bar1');
$r->exec();

/* Second multi block, still pipelined */
$r->multi();
$r->set('foo2', 'bar2');
$r->exec();

/* Send everything */
$r->exec();
```

Since we maintain the state (atomic, multi) in `redis_sock` we can also detect improper usage, like:

```php
$r->pipeline();
$r->pipeline(); // ignore or throw a warning?

$r->pipeline();
$r->multi();
$r->multi(); // warning or error?
```

I haven't looked at the PR in a while but I see no problem in only allowing `pipeline` as the outermost block.  I don't see any benefit in `multi() ... pipeline() ... exec() ... exec()` I think it would just confuse people.

**Comment by yatsukhnenko** [comment]
@michael-grunder 
```
$r->pipeline()->set('x', 1)->multi()->set('y', 1)->set('z', 1)->discard()
```
Should the `discard` to roll back changes of internal `multi`?

**Comment by michael-grunder** [comment]
That's an interesting question.  I suppose we _could_ do that but it seems like it would add a great deal of complexity.  The problem is that we would have to unwind the not yet sent commands (in the context of a pipeline).

I can imagine something like this happening:

```php
$r->pipeline();

$r->multi();
$r->set('k1','v1');
$r->set('k2','v2');
$r->discard();

$r->multi();
$r->set('k3','v3');
$r->set('k4','v4');
$r->exec();

$r->exec();
```

I'm not against unrolling the first multi block (when we encounter the discard command) but not sure if it's worth the added complexity that it would require.

So I don't have a good answer.  Do you think this is something we should do?

**Comment by yatsukhnenko** [comment]
We don't have an ability (if I didn't miss something) to discard commands that were added into pipeline, so I think that we should do this and we can put this logic into `discard` method. To do it less complexity, if we are in pipeline mode, we may discard ALL commands and don't think about kind of them. What do you think?

**Comment by yatsukhnenko** [comment]
@michael-grunder, we may land this (after your reviewing and adding tests) and separately implement `discard` if somebody will need this 😀

**Comment by michael-grunder** [comment]
Hey,

I did a little bit of testing on the branch and it seems to work for `pipeline()->multi()->...->exec()->exec()` but doesn't work for `multi()->pipeline()...exec()->exec()`.  I actually think this is good (it doesn't make sense to me to allow it both ways), but we should probably throw an error if the user attempts to enter a pipeline when already in multi mode.

So as it works now, discard in a pipeline will discard the _entire_ set of pipeline commands.  This is probably OK most of the time, but I am certain someone will run into that issue (expecting it to just discard the most recent multi attempt) and there will be bug reports :smile:

Would you mind creating a new branch from this repo's develop branch and then merging in these changes so that it's easier for both of us to work on them?

Mike

**Comment by yatsukhnenko** [comment]
@michael-grunder, I've created `multi-pipeline` branch in this repo

**Comment by zzjin** [comment]
any update here pls?

**Comment by yatsukhnenko** [comment]
@zzjin, all tests passed, you may try to use it

**Comment by zzjin** [comment]
seems all right, no error or warnings.

**Comment by ovr** [comment]
ping @michael-grunder

**Comment by michael-grunder** [comment]
I'll try to take a look this evening.

File: redis.c [hunk 1]
```diff
@@ -1242,9 +1242,11 @@ PHP_METHOD(Redis,__destruct) {
 
     // If we think we're in MULTI mode, send a discard
     IF_MULTI() {
-        // Discard any multi commands, and free any callbacks that have been
-        // queued
-        redis_send_discard(redis_sock TSRMLS_CC);
+        IF_NOT_PIPELINE() {
+            // Discard any multi commands, and free any callbacks that have been
+            // queued
+            send_discard_static(redis_sock TSRMLS_CC);
```
Comment (michael-grunder) [inline redis.c]: I think this needs to be changed to `redis_send_discard`

Comment (yatsukhnenko) [inline redis.c]: Nice catch! Was mistaken when rebasing

**Comment by michael-grunder** [comment]
I haven't done extensive testing but it sure seems to work for me locally.  I think it's fine to merge this into develop so that a few more people (who run bleeding edge) can let us know if it causes problems.

People are going to love this feature.

**Comment by zzjin** [comment]
May be there is one error here:
`symbol lookup error: php/lib/php/extensions/no-debug-non-zts-20160303/redis.so: undefined symbol: send_discard_static`
when error occr under multi state:
```php
$redis -> multi();
$redis -> del(test);
$redis -> hMSet('test',$a); //here $a is not defined or `NULL`
$redis -> exec();
```

But U have just merged 5mins ago...

**Comment by yatsukhnenko** [comment]
@zzjin, I've fixed this before merging

**Comment by zzjin** [comment]
@yatsukhnenko yes after update to latest `develop` branch 阿里
 seemd right. Thx~

**Pull request merged**